otri/validation/__init__.py

```diff
@@ -0,0 +1,572 @@
+        '''
+        self._add_label(data, result)
+        if self._holding[index]:
+            self._hold_buffer[index].append(data)
+        else:
+            self._push_data(data, index)
+
+    def _label_all(self, label: Any):
+        '''
+        Label all the atoms in the internal buffer.
+
+        Parameters:
+            label : Any
+                The label to append.
+        '''
+        for buffer in self._hold_buffer:
+            for atom in buffer:
+                self._add_label(atom, label)
+
+    def _on_inputs_closed(self):
+        '''
+        When the inputs are empty and closed everything gets released before closing the outputs.
+        '''
+        self._release_all()
+        return super()._on_inputs_closed()
+
+
+class ParallelValidator(ValidatorFilter, ParallelFilter):
+
+    '''
+    This filter handles finding errors between multiple Streams, in parallel,
+    popping one atom from each of them and consulting them together.
+
+    If a value is returned by _check, all parallel atoms are considered affected by it and get
+    labeled.
+    '''
+
+    def __init__(self, inputs: Sequence[str], outputs: Sequence[str], check: Callable = None):
+        '''
+        This Validator expects the same number of Stream inputs and outputs.
+
+        Parameters:
+            inputs : str
+                Names of the inputs.
+
+            outputs : str
+                Names of the outputs.
+
+            check : Callable
+                If you don't want to override the class, you can pass a Callable here.
+                The Callable should require the atom batch as a parameter.
+        '''
+        ParallelFilter.__init__(self, inputs, outputs)
+
+        if check is not None:
+            self._check = check
+
+    def _on_ok(self, data: List[Mapping], indexes: List[int]):
+        '''
+        Called if `_check` returned `None`.
+        Pushes the atom to the output on the same index it came from.
+
+        Parameters:
+            data : List[Mapping]
+                The checked data.
+
+            index : List[int]
+                The indexes from where each atom came from.
+        '''
+        for atom, index in zip(data, indexes):
+            self._push_data(atom, index)
+
+    def _on_result(self, data: List[Mapping], result: Any, indexes: List[int]):
+        '''
+        Called if `_check` returned something.
+        Labels all the atoms in the `data` list and pushes them all.
+
+        Parameters:
+            data : List[Mapping]
+                The checked data.
+
+            result : Any
+                The result of `_check`.
+
+            indexes : List[int]
+                The index of the input the data has been popped from.
+        '''
+        for atom, index in zip(data, indexes):
+            append_label(atom, result)
+            self._push_data(atom, index)
+
+    def _check(self, data: List[Mapping], indexes: List[int]):
```

Comment:
Codacy found an issue: [An attribute defined in otri.validation line 383 hides this method](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6277039)